server/service/translator.go

```diff
@@ -0,0 +1,113 @@
+package service
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kithttp "github.com/go-kit/kit/transport/http"
+)
+
+type translatorRequest struct {
+	List []fleet.TranslatePayload `json:"list"`
+}
+
+type translatorResponse struct {
+	List []fleet.TranslatePayload `json:"list"`
+	Err  error                    `json:"error,omitempty"`
+}
+
+func (r translatorResponse) error() error { return r.Err }
+
+func makeTranslatorEndpoint(svc fleet.Service, opts []kithttp.ServerOption) http.Handler {
+	return newServer(
+		makeAuthenticatedServiceEndpoint(svc, translatorEndpoint),
+		makeDecoderForType(translatorRequest{}),
+		opts,
+	)
+}
+
+func translatorEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*translatorRequest)
+	resp, err := svc.Translate(ctx, req.List)
+	if err != nil {
+		return translatorResponse{Err: err}, nil
+	}
+	return translatorResponse{List: resp}, nil
+}
+
+func (svc Service) Translate(ctx context.Context, payloads []fleet.TranslatePayload) ([]fleet.TranslatePayload, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.User{}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	var finalPayload []fleet.TranslatePayload
+
+	for _, payload := range payloads {
+		var toIDFunc func(string) (interface{}, error)
+		var idExtractorFunc func(interface{}) uint
+
+		switch payload.Type {
+		case fleet.TranslatorTypeUserEmail:
+			toIDFunc = func(email string) (interface{}, error) { return svc.ds.UserByEmail(email) }
+			idExtractorFunc = func(item interface{}) uint { return item.(*fleet.User).ID }
```

Comment:
Shall we create an interface for types with a datastore ID? Could help to improve type safety. On the other hand it would require implementing for each type (but looks like we will have to do some implementation for each type anyway).

Reply:
Sure! I'll do a pass with an interface instead, it'll probably be more readable.